### ISSUE TITLE
Matchable documentation

### ISSUE THREAD

**Issue by allquantor** [aol/cyclops#535]
Hi,
there is really hard to follow the examples in the documentation. Here is an example on Matchables:

The given documentation [here](https://github.com/aol/cyclops-react/blob/7466771e9ef93c7e028ba53d6ee652b7aea49208/user-guide/pattern-matching.adoc)
seems to be not up to date anymore since the package structure does not "match" :trollface: anymore to the last version in maven (2.0.0-MI4).

In detail, the part:

```
import static com.aol.cyclops2.control.Matchable.otherwise;
import static com.aol.cyclops2.control.Matchable.then;
import static com.aol.cyclops2.control.Matchable.when;

String m = Matchable.of(Optional.of(3))
                    .matches(
                         c->c.is(when(3),then("three")),otherwise("not three")
                        )
```

Is not possible because of missing control.Matchable in cyclops2. 
There is the same case with the Monatic documentation at least. 

is there any plan to consolidate the documentation? It would be really helpful!

**Comment by johnmcclean-aol** [comment]
hey @allquantor good point. Matchable is currently gone from 2.x, there is an 2.x proposal logged to further simplify pattern matching ( https://github.com/aol/cyclops-react/issues/422 ). I think in practice Matchable too complex for everyday use, but could be made a standalone module in cyclops. Would love to get your feedback on that?

The documentation should be updated to signify this is currently only available in 1.x.

2.x currently supports pattern matching via the visit method on each cyclops-react data type.

**Comment by allquantor** [comment]
@johnmcclean-aol Thanks for the fast response! This sounds great, seriously, decent pattern matching and higher kinded types are the things I miss most in Java, so your project gave me hope!

A really cool thing too would be a gist how to "idiomatically" use the typeclasses in cyclops. Tbh. I tried to implement one but gave up after being overwhelmed with the amount of additional types and interfaces I had to look into. I think that would make the project for people with fp background but not experienced more accessible.

**Comment by johnmcclean-aol** [comment]
@allquantor Sure - that sounds like a good topic for a blog post. 

The easiest way may actually be just to implement Functor, Monad, Pure etc. depending on how your own classes are set up. Functor only requires that you implement map and it takes a transformation function and a Higher Kind encoded type to transform. On your own classes you can just implement Higher (rather than create a proxy or wrapping type like we have to do for JDK or 3rd Party Types). 

Internally I've used the classes inside General (https://github.com/aol/cyclops-react/blob/master/src/main/java/cyclops/typeclasses/instances/General.java). Most of those instances contain method references to the actual transformation function, so the ListFunctor makes use of GeneralFunctor and (indirectly) passes in a reference to the map method on ListX, but like I said just creating your own Functor implementation might be simpler.

Here is an example Functor used in the test for Free (https://github.com/aol/cyclops-react/blob/master/src/test/java/cyclops/free/CharToy.java#L41)

**Comment by johnmcclean-aol** [comment]
Also this project might be interesting to you https://github.com/derive4j/hkt, it uses a preprocessor to enforce HKT encodings . I want to make use of it in cyclops but at the moment I'm getting compile errors in IntelliJ when I switch it on, when I get some time I'll circle back to the derive4j team.

**Comment by johnmcclean-aol** [comment]
In case you want to use the typeclasses that come bundled with cyclops-react, each datatype typically has a nested Instances class. So we can just write

```java
     ListX.Instances.functor().map(i->i+1,ListX.of(10,20,30));
```

**Comment by johnmcclean** [comment]
Need to document the new Pattern Matching API
